python/pacemaker/_cts/watcher.py

```diff
@@ -229,33 +231,33 @@ def async_complete(self, pid, returncode, out, err):
         out         -- stdout from the journal read
         err         -- stderr from the journal read
         """
-        found_cursor = False
-        for line in out:
-            match = re.search(r"^-- cursor: ([^.]+)", line)
+        parser = isoparser()
+        dt_limit = None
+        if self.limit:
+            dt_limit = parser.isoparse(self.limit)
 
+        for line_num, line in enumerate(out):
+            match = re.search(r"^-- cursor: ([^.]+)", line)
             if match:
-                found_cursor = True
                 self.offset = match.group(1).strip()
-                self.debug("Got %d lines, new cursor: %s" % (len(out), self.offset))
-            else:
-                self.cache.append(line)
+                self.debug("Got %d lines before new cursor: %s"
+                           % (line_num, self.offset))
+                break
 
-        if self.limit and not found_cursor:
```

Comment:
I don't think this was possible. `journalctl ... --show-cursor` always shows a cursor, even if the output is otherwise empty because there are no more logs between the old cursor and the `--until` timestamp. So when we parse the output in the `for` loop, we'll find a cursor.

If anyone's curious, then we could test on a RHEL cluster. My VMs are all Fedora right now, where `LogWatcher.look()` loops infinitely (the thing we're fixing here).

Reply:
Watching logs still seems subtly buggy.

The new approach no longer updates the cursor if we hit the limit. That's a problem only if we `look()` again with the same `LogWatcher` object. We may fetch and parse the same batch of logs twice.

`LogWatcher.set_watch()` initializes the `JournalObj` object's cursor. This is before any timeout starts counting down or the `limit` is set.

Now, when `LogWatcher.look()` returns:
* If we **didn't** hit the limit, then the cursor is updated to the end of the batch of lines we fetched. Any unused lines remain in `LogWatcher._line_cache` to be read before another `LogWatcher.__get_lines()`. The next `LogWatcher.__get_lines()` call will start with the next batch of lines. Everything should be okay.
* If we **did** hit the limit, then the cursor remains at the beginning of this batch of lines. We process the lines between the cursor and the limit timestamp, and any unused ones stay in the cache. However, we'll fetch the same lines again if we call `look()` again. Two ways to avoid this:
    * Make another `journalctl` call (probably using `--until <limit> --lines 0` or `--after-cursor=<old_cursor> --lines=<lines-until-limit-was-found>`) to get the new cursor after we hit the limit.
    * Check timestamps against the limit within `LogWatcher.look()`, instead of within `JournalObj.async_complete()`. This avoids another `journalctl` call but it's not ideal design because it means tighter coupling between the classes.

Reply:
I'll work on this on Monday. Plan is to maintain log messages from after the "until" timestamp, in the `JournalObj`'s cache (not in the `LogWatcher` line cache), but it's gonna require some shuffling around.